Rework group page routing/state to support back

nicksellen opened the pull request:
Not finished yet, there are 6 places that use the `GroupSelect` component that need to be reworked, I did 2 so far.

It was a bit of tangle of state with the values being read/set in both the page component and the `GroupSelect` component, was quite hard to reason about the ordering. The tricky bit with that architecture was there was no real way to tell from the page component when the group had changed because someone had actively selected one, or it was updating from the store on initial load.

So, I thought to make the `GroupSelect` more dumb. Now, it gets the list of groups from the store, but does not handle the remembering group feature. That is handled by a new mixin, to keep the page components simple too.

So, the new responsibility of the `GroupSelect` is just to implement a `v-model` interface for setting/getting the value of the group.

The mixin is where the brains are now, in one place, it's job is to check the route params, check the remember setting, and bind the correct group value for the `GroupSelect` to use.

It's a bit fiddly as there are a surprisingly large number of value/state combinations, but it's at least in one place! (was previously spread between each page component and the `GroupSelect` component.

I also added a `group/forget` action - so that when you select "all groups" (id=0) it doesn't save that, but clears the state. I think that more accurately reflects the intention.

Also, it's important to `unwatch()` watchers if they are created in a component, see https://forum.vuejs.org/t/vuex-unwatch-getter-is-this-right/68154/4 for details.

I'll add a few comments inline to explain how it works....

I would also quite like to add a test to recreate what I've been doing manually for now, as it could be easy to accidentality break.

nicksellen commented on components/GroupSelect.vue:
```diff
@@ -11,37 +11,29 @@ select {
 <script>
 export default {
   props: {
+    value: {
```
This is how to implement a `v-model` inside a custom component, this and doing `$emit('input', newVal)` elsewhere.

nicksellen commented on mixins/createGroupRouteMixin.js:
```diff
@@ -0,0 +1,69 @@
+ * @param key used for the remember key and the route segment
+ * @returns a Vue mixin
+ */
+export default function createGroupRouteMixin(key) {
```
At the moment assumes the remember key and the route path are the same, e.g. `communityevents` is the remember key and `/communityevents/` is the path, but could added more args to implement that if needed.

nicksellen commented on mixins/createGroupRouteMixin.js:
```diff
@@ -0,0 +1,69 @@
+  }
+  return {
+    computed: {
+      groupid: {
```
Anything using this mixin automatically gets a readable/writable `groupid` field.

nicksellen commented on mixins/createGroupRouteMixin.js:
```diff
@@ -0,0 +1,69 @@
+      }
+    },
+    beforeDestroy() {
+      if (this._unwatchGroupRemember) this._unwatchGroupRemember()
```
Important! Or old instances get left around wrecking havoc and confusion!

See https://forum.vuejs.org/t/vuex-unwatch-getter-is-this-right/68154/4

nicksellen commented on mixins/createGroupRouteMixin.js:
```diff
@@ -0,0 +1,69 @@
+      this._unwatchGroupRemember = this.$store.watch(
+        (state, getters) => getters['group/remembered'](rememberId),
+        (val, oldVal) => {
+          if (oldVal === undefined && val === undefined && this.groupid !== 0) {
```
This logic is a bit gnarly, and hard to understand I think....

Due to using `immediate: true` below, we get called _immediately_ (clever naming huh!) with the values, so we don't need to make a seperate call to get the remember value first....

nicksellen commented on mixins/createGroupRouteMixin.js:
```diff
@@ -0,0 +1,69 @@
+            if (this.groupid === 0) {
+              // We have a remember value, but we're on the general page
+              // Replace the current route
+              this.$router.replace(routePath(val))
```
This is the magic method to change the route without adding it into the back stack....

nicksellen commented on pages/communityevents/_id.vue:
```diff
@@ -5,10 +5,11 @@
       <b-col cols="12" md="6" class="p-0">
         <div>
           <h1>Community Events</h1>
+          <pre>groupid: {{ groupid }}</pre>
```
Whoops, need to remove this.

nicksellen commented on pages/communityevents/_id.vue:
```diff
@@ -63,31 +65,10 @@ export default {
   },
-
   methods: {
-    groupChange: function(newGroup) {
```
This page component gets much simpler too, with all this logic outsourced to the mixin...

nicksellen commented:
Ok, thats me for today, there is a nice new component for wrapping the group select with memory.

Next bit is going through the remaining few places with a group select that needs either replacing with the wrapper wrapper component (volunteer and event modals), or doing the routing thing (stories)

nicksellen commented on components/CommunityEventModal.vue:
```diff
@@ -101,7 +101,7 @@
               For which community?
             </label>
-            <groupSelect id="editevent" class="" @change="groupChange" />
+            <groupRememberSelect v-model="groupid" remember="editevent" />
```
Not sure if this works correctly when editing an existing event. I created one in the playground group, but it disappeared after pressing save, maybe into moderation?

edwh replied:
Yes, and I later zapped it.  Create a new one if you want and I'll approve it.

nicksellen commented on components/CommunityEventModal.vue:
```diff
@@ -326,17 +326,6 @@ export default {
 
       // If we don't have any groups, force a select.
       this.event.groups = this.event.groups ? this.event.groups : [{ id: 0 }]
```
Maybe this line can be removed too, I didn't quite understand why it's there...

edwh replied:
When we first add an event, we won't have any groups.  We want to force them to choose one.  That might happen because of the order of the select options (as choose as at the top) but I see no harm in making it explicit.

edwh replied:
We have also seen cases in the past where events get stranded without groups due to bugs and we want to edit them to fix them up.

nicksellen replied:
I still don't really understand. It seems an event can have multiple groups (as I see the AddGroup/RemoveGroup stuff) but the UI only has a place for setting a single group?

![singlegroup](https://user-images.githubusercontent.com/31616/68590953-faa25b00-048f-11ea-8861-5a928c65b624.png)

nicksellen commented on mixins/createGroupRouteMixin.js:
```diff
@@ -0,0 +1,89 @@
+        }
+      }
+    },
+    created() {
```
TODO: check how this is for server side stuff, as `created` is also called on the server...

edwh commented:
Thanks for this.
GroupSelect - Can you add a TODO DESIGN for the max-width setting in 
GroupSelect.  I think I put that hardcoded width in there but it can't 
be right.createGroupRouteMixinHas quite a lot of hardcoded 0s.  Constant 
for clarity?Sorry, nitpicking, but I'm not mad on the name having 
"create" and "mixin" in there.  The mixins/folder tells you it's a 
mixin, and all mixins would qualify to be called create, wouldn't they?  
For my education: When is it correct to set up a watcher inside 
created() rather than use the watch: option 
(https://vuejs.org/v2/guide/computed.html#Watchers)?  I'd rashly assumed 
that took care of destruction for you, but I'm probably missing 
something.Based on the previous point, can you (if you haven't already) 
check for any other watch leaks?StartEndCollection We removed all uses 
of index in keys because they are bad - see discussion with Jason.  But 
you've put this one back in explicitly.  Why's that?

------ Original Message ------
From: "Nick Sellen" <notifications@github.com>
To: "Freegle/iznik-nuxt" <iznik-nuxt@noreply.github.com>
Cc: "Subscribed" <subscribed@noreply.github.com>
Sent: 08/11/2019 11:18:27
Subject: Re: [Freegle/iznik-nuxt] Rework group page routing/state to 
support back (#33)

>@nicksellen commented on this pull request.
>
>--------------------------------------------------------------------------------
>In components/CommunityEventModal.vue 
><https://github.com/Freegle/iznik-nuxt/pull/33#discussion_r344128769>:
>
> > @@ -326,17 +326,6 @@ export default {
>
>        // If we don't have any groups, force a select.
>        this.event.groups = this.event.groups ? this.event.groups : [{ id: 0 }]
>
>Maybe this line can be removed too, I didn't quite understand why it's 
>there...
>
>—
>You are receiving this because you are subscribed to this thread.
>Reply to this email directly, view it on GitHub 
><https://github.com/Freegle/iznik-nuxt/pull/33?email_source=notifications&email_token=AAAN26F3IC2CGNXOFUJA6OTQSVDIHA5CNFSM4JKGW5E2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCK4WPMY#pullrequestreview-314140595>, 
>or unsubscribe 
><https://github.com/notifications/unsubscribe-auth/AAAN26FLJ4JAOKDPDNWBSADQSVDIHANCNFSM4JKGW5EQ>.
>

nicksellen commented:
> Can you add a TODO DESIGN for the max-width setting in GroupSelect.

Yup!

> Sorry, nitpicking, but I'm not mad on the name having "create" and "mixin" in there.

Agree about the "*Mixin" suffix, will change that. This one returns a function that creates a mixin (to enable parameters to be passed in) rather than a mixin directly hence the _create_ prefix, makes it clearer that you need to call it first, not just use it directly. What do you think?

> createGroupRouteMixinHas quite a lot of hardcoded 0s.  Constant for clarity?

Yup, will do. would name it `ALL_GROUPS` or something... 0 was already used for that, so I hope there's not actually a group with id=0.

> For my education: When is it correct to set up a watcher inside created() rather than use the watch: option (https://vuejs.org/v2/guide/computed.html#Watchers)?  I'd rashly assumed that took care of destruction for you, but I'm probably missing something.

Using `watch:` option, the destruction is taken care of. You use that option to watch properties that are on your vue instance, in this case we're watching a getter on the store, so can't put the getter expression as a key in the `watch`: option. Having said that, you can declare a computed property that uses the getter, then watch that... that's what I used in the GroupRememberSelect iirc.

Something like:

```js
{
  computed: {
    foo() {
      return this.$store.getters['blah']
    }
  },
  watch: {
    foo(val) {
      console.log('foo changed!', val)
    }
  }
}
```

The more proper way actually for this is to use [mapGetters](https://vuex.vuejs.org/guide/getters.html#the-mapgetters-helper) from vuex to bring the getter value into a local property, but you have defined all the getters as functions, so can't use that. That's another topic for now ;)

```js
{
  computed: {
    ...mapGetters({
      foo: 'somemodule/somegetter'
    })
  }
}
```

> Based on the previous point, can you (if you haven't already) check for any other watch leaks?

Would do that in a seperate PR to keep this one focused.

> StartEndCollection We removed all uses of index in keys because they are bad - see discussion with Jason.  But you've put this one back in explicitly.  Why's that?

The replacement key was totally wrong as `date.toISOString` is not a function, it's an object with a start/end field, that can both be null, so it's not a valid key at all. I just reverted the change to avoid unrelated changes here, I don't know what it should use for the key.

edwh commented:
------ Original Message ------
From: "Nick Sellen" <notifications@github.com>
To: "Freegle/iznik-nuxt" <iznik-nuxt@noreply.github.com>
Cc: "Edward Hibbert" <edward@ehibbert.org.uk>; "Review requested" 
<review_requested@noreply.github.com>
Sent: 08/11/2019 15:17:48
Subject: Re: [Freegle/iznik-nuxt] Rework group page routing/state to 
support back (#33)

>>Can you add a TODO DESIGN for the max-width setting in GroupSelect.
>>
>Yup!
>
>>Sorry, nitpicking, but I'm not mad on the name having "create" and 
>>"mixin" in there.
>>
>Agree about the "*Mixin" suffix, will change that. This one returns a 
>function that creates a mixin (to enable parameters to be passed in) 
>rather than a mixin directly hence the create prefix, makes it clearer 
>that you need to call it first, not just use it directly. What do you 
>think?
>
Ah yes.  Fine.
>
>
>>createGroupRouteMixinHas quite a lot of hardcoded 0s. Constant for 
>>clarity?
>>
>Yup, will do. would name it ALL_GROUPS or something... 0 was already 
>used for that, so I hope there's not actually a group with id=0.
>
There isn't.
>
>
>>For my education: When is it correct to set up a watcher inside 
>>created() rather than use the watch: option 
>>(https://vuejs.org/v2/guide/computed.html#Watchers)? I'd rashly 
>>assumed that took care of destruction for you, but I'm probably 
>>missing something.
>>
>Using watch: option, the destruction is taken care of. You use that 
>option to watch properties that are on your vue instance, in this case 
>we're watching a getter on the store, so can't put the getter 
>expression as a key in the watch: option. Having said that, you can 
>declare a computed property that uses the getter, then watch that... 
>that's what I used in the GroupRememberSelect iirc.
>
>Something like:
>
>{
>   computed: {
>     foo() {
>       returnthis.$store.getters['blah']
>     }
>   },
>   watch: {
>     foo(val) {
>       console.log('foo changed!', val)
>     }
>   }
>}
>The more proper way actually for this is to use mapGetters 
><https://vuex.vuejs.org/guide/getters.html#the-mapgetters-helper> from 
>vuex to bring the getter value into a local property, but you have 
>defined all the getters as functions, so can't use that. That's another 
>topic for now ;)
>
>{
>   computed: {
>     ...mapGetters({
>       foo:'somemodule/somegetter'
>     })
>   }
>}
Ta.
>
>>Based on the previous point, can you (if you haven't already) check 
>>for any other watch leaks?
>>
>Would do that in a seperate PR to keep this one focused.
>
Ok.
>
>
>>StartEndCollection We removed all uses of index in keys because they 
>>are bad - see discussion with Jason. But you've put this one back in 
>>explicitly. Why's that?
>>
>The replacement key was totally wrong as date.toISOString is not a 
>function, it's an object with a start/end field, that can both be null, 
>so it's not a valid key at all. I just reverted the change to avoid 
>unrelated changes here, I don't know what it should use for the key.
>
I see.  Just JSON encode it, that would work, I think?
>
>
>—
>You are receiving this because your review was requested.
>Reply to this email directly, view it on GitHub 
><https://github.com/Freegle/iznik-nuxt/pull/33?email_source=notifications&email_token=AAAN26DBJID3GO3QWVMIPUTQSV7JZA5CNFSM4JKGW5E2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEDSNKNI#issuecomment-551867701>, 
>or unsubscribe 
><https://github.com/notifications/unsubscribe-auth/AAAN26D6FNH5OAVS57MVS63QSV7JZANCNFSM4JKGW5EQ>.
>

nicksellen commented:
> I see.  Just JSON encode it, that would work, I think?

That wouldn't work. They key is to describe a persistent identity of the item. Using the JSON encoded value would use the value as identity - but the value will change when you select date/time values.

I think it is broken using id (when I press remove and have two dates listed, the wrong one is going...), but even more broken when using "date.toISOString". Needs more thought/work than I think is appropriate here... which is actually meant to just be about the back button ;)

edwh commented:
You're right - mission creep.  Stick a TODO in there as part of your fix 
so we don't lose it.

------ Original Message ------
From: "Nick Sellen" <notifications@github.com>
To: "Freegle/iznik-nuxt" <iznik-nuxt@noreply.github.com>
Cc: "Edward Hibbert" <edward@ehibbert.org.uk>; "Review requested" 
<review_requested@noreply.github.com>
Sent: 08/11/2019 15:31:05
Subject: Re: [Freegle/iznik-nuxt] Rework group page routing/state to 
support back (#33)

>>I see. Just JSON encode it, that would work, I think?
>>
>That wouldn't work. They key is to describe a persistent identity of 
>the item. Using the JSON encoded value would use the value as identity 
>- but the value will change when you select date/time values.
>
>I think it is broken using id (when I press remove and have two dates 
>listed, the wrong one is going...), but even more broken when using 
>"date.toISOString". Needs more thought/work than I think is appropriate 
>here... which is actually meant to just be about the back button ;)
>
>—
>You are receiving this because your review was requested.
>Reply to this email directly, view it on GitHub 
><https://github.com/Freegle/iznik-nuxt/pull/33?email_source=notifications&email_token=AAAN26GIZUZLECWTRKJ7X73QSWA3RA5CNFSM4JKGW5E2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEDSOSKI#issuecomment-551872809>, 
>or unsubscribe 
><https://github.com/notifications/unsubscribe-auth/AAAN26G37MITDNLOAN2QWNTQSWA3RANCNFSM4JKGW5EQ>.
>

edwh commented:
Half-baked future-proofing.  For messages, events, volunteer ops, they 
can all only have a single group at present.  But they may have multiple 
groups in future.

------ Original Message ------
From: "Nick Sellen" <notifications@github.com>
To: "Freegle/iznik-nuxt" <iznik-nuxt@noreply.github.com>
Cc: "Edward Hibbert" <edward@ehibbert.org.uk>; "State change" 
<state_change@noreply.github.com>
Sent: 11/11/2019 13:35:36
Subject: Re: [Freegle/iznik-nuxt] Rework group page routing/state to 
support back (#33)

>@nicksellen commented on this pull request.
>
>--------------------------------------------------------------------------------
>In components/CommunityEventModal.vue 
><https://github.com/Freegle/iznik-nuxt/pull/33#discussion_r344713040>:
>
> > @@ -326,17 +326,6 @@ export default {
>
>        // If we don't have any groups, force a select.
>        this.event.groups = this.event.groups ? this.event.groups : [{ id: 0 }]
>
>I still don't really understand. It seems an event can have multiple 
>groups (as I see the AddGroup/RemoveGroup stuff) but the UI only has a 
>place for setting a single group?
>
>singlegroup
><https://user-images.githubusercontent.com/31616/68590953-faa25b00-048f-11ea-8861-5a928c65b624.png>
>
>—
>You are receiving this because you modified the open/close state.
>Reply to this email directly, view it on GitHub 
><https://github.com/Freegle/iznik-nuxt/pull/33?email_source=notifications&email_token=AAAN26DORZP33SKRXJVZ7NLQTFNSRA5CNFSM4JKGW5E2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCLCPWTY#discussion_r344713040>, 
>or unsubscribe 
><https://github.com/notifications/unsubscribe-auth/AAAN26BSKUC35MZSQYWVJNTQTFNSRANCNFSM4JKGW5EQ>.
>
